features/FEATURE_LWIP/lwip-interface/lwip_stack.c

```diff
@@ -1023,6 +1043,35 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
             }
             return 0;
 
+        case NSAPI_ADD_MEMBERSHIP:
+        case NSAPI_DROP_MEMBERSHIP: {
+#if LWIP_IPV4
+            if (optlen != sizeof(nsapi_ip_mreq_t)) {
+                return NSAPI_ERROR_UNSUPPORTED;
+            }
+
+            err_t igmp_err;
+            nsapi_ip_mreq_t *imr = optval;
+
+            ip_addr_t if_addr;
+            ip_addr_t multi_addr;
+
+            nsapi_addr_to_ip_addr(&multi_addr, &imr->imr_multiaddr);
+            nsapi_addr_to_ip_addr(&if_addr, &imr->imr_interface);
```

Comment:
For full IPv4/IPv6 correctness, I think you need to make sure that if multi_addr.version is NSAPI_UNSPEC, you produce an ip_addr_t of the correct type using ip_addr_set_zero_ip4 or ip_addr_set_zero_ip6.

I think I'd restructure like:

    /* Check interface address type matches group, or is unspecified */
    if (imr->imr_interface.version != NSAPI_UNSPEC && imr->imr_interface.version != imr->imr_multiaddr.version) {
        return NSAPI_ERROR_PARAMETER;
    }

    ip_addr_t if_addr;
    ip_addr_t multi_addr;

    if (!convert_mbed_addr_to_lwip(&multi_addr, &imr->imr_multiaddr)) {
        return NSAPI_ERROR_PARAMETER;
    }
    if (imr->imr_interface.version != NSAPI_UNSPEC) {
        if (!convert_mbed_addr_to_lwip(&if_addr, &imr->imr_interface)) {
            return NSAPI_ERROR_PARAMETER;
        }
    } else {
        ip_addr_set_any(IP_IS_V6(if_addr), &imr->imr_interface)
    }

    igmp_err = NSAPI_ERROR_UNSUPPORTED;
    #if LWIP_IPV4
    if (IP_IS_V4(if_addr)) {
        igmp_err = (NSAPI_ADD_MEMBERSHIP ?  igmp_joingroup : igmp_leavegroup)
                                (ip_2_ip4(&if_addr), ip_2_ip4(&multi_addr));
    }
    #endif 
    #if LWIP_IPV6
    if (IP_IS_V6(if_addr)) {
        igmp_err = (NSAPI_ADD_MEMBERSHIP ? mld6_joingroup : mld6_leavegroup)
                                (ip_2_ip6(&if_addr), ip_2_ip6(&multi_addr));
    }
    #endif

Untested, but I think should cover all the compilation/runtime permutations of IPv4/IPv6.